File: iOS/KinKitchen/KinKitchen/Features/Profile/EditProfileView.swift
//
//  EditProfileView.swift
//  KinKitchen
//
//  Created by Greg Hudler on 8/27/26.
//

import SwiftUI

struct EditProfileView: View {

    @Environment(\.dismiss) private var dismiss

    @State private var firstName = ""
    @State private var lastName = ""
    @State private var displayName = ""
    @State private var username = ""
    @State private var location = ""
    @State private var birthDate = Date()
    @State private var bio = ""

    @State private var errorMessage: String?
    @State private var isLoading = true
    @State private var isSaving = false

    var body: some View {
        ScrollView {
            VStack(spacing: KinSpacing.xLarge) {

                Text("Edit Profile")
                    .font(KinTypography.largeTitle)
                    .foregroundStyle(KinColors.primaryText)

                if isLoading {
                    ProgressView()

                } else {
                    profileForm
                }
            }
            .padding(KinSpacing.xLarge)
        }
        .background(KinColors.background)
        .task {
            await loadProfile()
        }
    }

    private var profileForm: some View {
        VStack(spacing: KinSpacing.xLarge) {

            VStack(
                alignment: .leading,
                spacing: KinSpacing.large
            ) {
                KinTextField(
                    title: "First Name",
                    text: $firstName
                )

                KinTextField(
                    title: "Last Name",
                    text: $lastName
                )

                KinTextField(
                    title: "Display Name",
                    text: $displayName
                )

                KinTextField(
                    title: "Username",
                    text: $username
                )

                KinTextField(
                    title: "Location",
                    text: $location
                )

                VStack(
                    alignment: .leading,
                    spacing: KinSpacing.small
                ) {
                    Text("Birthday")
                        .font(KinTypography.body)
                        .foregroundStyle(KinColors.primaryText)

                    DatePicker(
                        "Birthday",
                        selection: $birthDate,
                        in: allowedBirthDateRange,
                        displayedComponents: .date
                    )
                    .labelsHidden()
                    .datePickerStyle(.compact)
                }

                VStack(
                    alignment: .leading,
                    spacing: KinSpacing.small
                ) {
                    Text("Bio")
                        .font(KinTypography.body)
                        .foregroundStyle(KinColors.primaryText)

                    KinTextEditor(
                        text: $bio
                    )
                }
            }

            if let errorMessage {
                Text(errorMessage)
                    .font(KinTypography.caption)
                    .foregroundStyle(KinColors.error)
                    .frame(
                        maxWidth: .infinity,
                        alignment: .leading
                    )
            }

            KinPrimaryButton(
                title: isSaving ? "Saving..." : "Save",
                color: KinColors.success
            ) {
                Task {
                    await saveProfile()
                }
            }
            .disabled(isSaving)

            KinSecondaryButton(
                title: "Cancel"
            ) {
                dismiss()
            }
        }
    }

    private var allowedBirthDateRange: ClosedRange<Date> {
        let calendar = Calendar.current

        let oldest =
            calendar.date(
                byAdding: .year,
                value: -120,
                to: Date()
            ) ?? Date()

        let youngest =
            calendar.date(
                byAdding: .year,
                value: -13,
                to: Date()
            ) ?? Date()

        return oldest...youngest
    }

    @MainActor
    private func loadProfile() async {
        isLoading = true
        errorMessage = nil

        defer {
            isLoading = false
        }

        do {
            let profile =
                try await ProfileService.fetchCurrentProfile()

            firstName = profile.firstName ?? ""
            lastName = profile.lastName ?? ""
            displayName = profile.displayName ?? ""
            username = profile.username ?? ""
            location = profile.location ?? ""
            bio = profile.bio ?? ""

            if let savedBirthDate =
                ProfileService.date(from: profile.birthDate)
            {
                birthDate = savedBirthDate
            }

        } catch {
            errorMessage = "Unable to load your profile."
        }
    }

    @MainActor
    private func saveProfile() async {
        guard !isSaving else {
            return
        }

        let cleanFirstName =
            firstName.trimmingCharacters(
                in: .whitespacesAndNewlines
            )

        let cleanLastName =
            lastName.trimmingCharacters(
                in: .whitespacesAndNewlines
            )

        let cleanDisplayName =
            displayName.trimmingCharacters(
                in: .whitespacesAndNewlines
            )

        let cleanUsername =
            username.trimmingCharacters(
                in: .whitespacesAndNewlines
            )

        guard !cleanFirstName.isEmpty else {
            errorMessage = "First name is required."
            return
        }

        guard !cleanLastName.isEmpty else {
            errorMessage = "Last name is required."
            return
        }

        guard !cleanDisplayName.isEmpty else {
            errorMessage = "Display name is required."
            return
        }

        guard !cleanUsername.isEmpty else {
            errorMessage = "Username is required."
            return
        }

        isSaving = true
        errorMessage = nil

        defer {
            isSaving = false
        }

        do {
            try await ProfileService.updateProfile(
                firstName: cleanFirstName,
                lastName: cleanLastName,
                displayName: cleanDisplayName,
                username: cleanUsername,
                location: location,
                birthDate: birthDate,
                bio: bio
            )

            dismiss()

        } catch {
            errorMessage =
                "Unable to save your profile."
        }
    }
}

#Preview {
    EditProfileView()
}

File: Android/app/src/main/java/com/pushtomaindev/kinkitchen/features/profile/EditProfileView.kt
package com.pushtomaindev.kinkitchen.features.profile

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.pushtomaindev.kinkitchen.components.buttons.KinPrimaryButton
import com.pushtomaindev.kinkitchen.components.buttons.KinSecondaryButton
import com.pushtomaindev.kinkitchen.components.feedback.KinLoadingView
import com.pushtomaindev.kinkitchen.components.inputs.KinTextEditor
import com.pushtomaindev.kinkitchen.components.inputs.KinTextField
import com.pushtomaindev.kinkitchen.services.supabase.ProfileService
import com.pushtomaindev.kinkitchen.ui.theme.*
import kotlinx.coroutines.launch
import kotlinx.datetime.*

private const val MIN_AGE = 13
private const val MAX_AGE = 120

/** Mirrors iOS `EditProfileView`. */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun EditProfileView(
    modifier: Modifier = Modifier,
    onDismiss: () -> Unit,
) {
    val scope = rememberCoroutineScope()
    val today = remember { Clock.System.todayIn(TimeZone.currentSystemDefault()) }

    // iOS constrains the picker to (today - 120y)...(today - 13y), so the age
    // rule is enforced by the picker rather than by a save-time check.
    val oldestAllowed = remember(today) { today.minus(MAX_AGE, DateTimeUnit.YEAR) }
    val youngestAllowed = remember(today) { today.minus(MIN_AGE, DateTimeUnit.YEAR) }

    var firstName by rememberSaveable { mutableStateOf("") }
    var lastName by rememberSaveable { mutableStateOf("") }
    var displayName by rememberSaveable { mutableStateOf("") }
    var username by rememberSaveable { mutableStateOf("") }
    var location by rememberSaveable { mutableStateOf("") }
    var bio by rememberSaveable { mutableStateOf("") }
    var birthDate by remember { mutableStateOf(youngestAllowed) }

    var errorMessage by remember { mutableStateOf<String?>(null) }
    var isLoading by remember { mutableStateOf(true) }
    var isSaving by remember { mutableStateOf(false) }
    var isShowingDatePicker by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) {
        isLoading = true
        errorMessage = null
        try {
            val profile = ProfileService.fetchCurrentProfile()
            firstName = profile.firstName.orEmpty()
            lastName = profile.lastName.orEmpty()
            displayName = profile.displayName.orEmpty()
            username = profile.username.orEmpty()
            location = profile.location.orEmpty()
            bio = profile.bio.orEmpty()
            ProfileService.birthDate(profile.birthDate)?.let { birthDate = it }
        } catch (e: Exception) {
            errorMessage = "Unable to load your profile."
        } finally {
            isLoading = false
        }
    }

    fun saveProfile() {
        if (isSaving) return

        val cleanFirstName = firstName.trim()
        val cleanLastName = lastName.trim()
        val cleanDisplayName = displayName.trim()
        val cleanUsername = username.trim()

        errorMessage = when {
            cleanFirstName.isEmpty() -> "First name is required."
            cleanLastName.isEmpty() -> "Last name is required."
            cleanDisplayName.isEmpty() -> "Display name is required."
            cleanUsername.isEmpty() -> "Username is required."
            else -> null
        }
        if (errorMessage != null) return

        scope.launch {
            isSaving = true
            try {
                ProfileService.updateProfile(
                    firstName = cleanFirstName,
                    lastName = cleanLastName,
                    displayName = cleanDisplayName,
                    username = cleanUsername,
                    location = location,
                    birthDate = birthDate,
                    bio = bio,
                )
                onDismiss()
            } catch (e: Exception) {
                errorMessage = "Unable to save your profile."
            } finally {
                isSaving = false
            }
        }
    }

    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(KinSpacing.xLarge),
        modifier = modifier
            .fillMaxSize()
            .background(KinColors.background)
            .verticalScroll(rememberScrollState())
            .padding(KinSpacing.xLarge),
    ) {
        Text("Edit Profile", style = KinTypography.largeTitle, color = KinColors.primaryText)

        if (isLoading) {
            Box(Modifier.fillMaxWidth().height(200.dp)) { KinLoadingView() }
        } else {
            Column(verticalArrangement = Arrangement.spacedBy(KinSpacing.large)) {
                KinTextField("First Name", firstName, { firstName = it })
                KinTextField("Last Name", lastName, { lastName = it })
                KinTextField("Display Name", displayName, { displayName = it })
                KinTextField("Username", username, { username = it })
                KinTextField("Location", location, { location = it })

                Column(verticalArrangement = Arrangement.spacedBy(KinSpacing.small)) {
                    Text("Birthday", style = KinTypography.body, color = KinColors.primaryText)
                    KinSecondaryButton(birthDate.longDisplayDate()) { isShowingDatePicker = true }
                }

                Column(verticalArrangement = Arrangement.spacedBy(KinSpacing.small)) {
                    Text("Bio", style = KinTypography.body, color = KinColors.primaryText)
                    KinTextEditor(bio, { bio = it })
                }
            }

            errorMessage?.let {
                Text(it, style = KinTypography.caption, color = KinColors.error)
            }

            KinPrimaryButton(
                title = if (isSaving) "Saving..." else "Save",
                color = KinColors.success,
                enabled = !isSaving,
            ) { saveProfile() }

            KinSecondaryButton("Cancel") { onDismiss() }
        }
    }

    if (isShowingDatePicker) {
        val state = rememberDatePickerState(
            initialSelectedDateMillis = birthDate.atStartOfDayIn(TimeZone.UTC).toEpochMilliseconds(),
            selectableDates = object : SelectableDates {
                override fun isSelectableDate(utcTimeMillis: Long): Boolean {
                    val date = Instant.fromEpochMilliseconds(utcTimeMillis)
                        .toLocalDateTime(TimeZone.UTC).date
                    return date in oldestAllowed..youngestAllowed
                }
            },
        )

        DatePickerDialog(
            onDismissRequest = { isShowingDatePicker = false },
            confirmButton = {
                TextButton(onClick = {
                    state.selectedDateMillis?.let {
                        birthDate = Instant.fromEpochMilliseconds(it)
                            .toLocalDateTime(TimeZone.UTC).date
                    }
                    isShowingDatePicker = false
                }) { Text("OK") }
            },
            dismissButton = {
                TextButton(onClick = { isShowingDatePicker = false }) { Text("Cancel") }
            },
        ) { DatePicker(state = state) }
    }
}

/** Matches iOS `.formatted(date: .long, time: .omitted)`. */
internal fun LocalDate.longDisplayDate(): String {
    val month = month.name.lowercase().replaceFirstChar { it.uppercase() }
    return "$month $dayOfMonth, $year"
}
